TransferMethod/Sources/View/Widget/TextWidget.swift

```diff
@@ -69,6 +89,183 @@ class TextWidget: AbstractWidget {
+            handleTextForLettersAndNumbers(currentText: currentText,
+                                           currentTextCharacter: filteredCharacter,
+                                           currentTextIndex: &currentTextIndex,
+                                           finalText: &finalText,
+                                           pattern: pattern,
+                                           patternCharactersToBeWritten: &patternCharactersToBeWritten,
+                                           patternIndex: &patternIndex)
+
+        default:
+            isEscapedCharacter = self.isEscapedCharacter(currentPatternCharacter.first!)
+            if !isEscapedCharacter {
+                if String(currentPatternCharacter) == currentTextCharacter {
+                    finalText += currentTextCharacter
+                    currentTextIndex = currentText.index(after: currentTextIndex)
+                } else {
+                    patternCharactersToBeWritten += currentPatternCharacter
+                }
+            }
+            patternIndex = pattern.index(after: patternIndex)
+        }
+    }
+
+    private func handleTextForLettersAndNumbers(currentText: String,
+                                                currentTextCharacter: String?,
+                                                currentTextIndex: inout String.Index,
+                                                finalText: inout String,
+                                                pattern: String,
+                                                patternCharactersToBeWritten: inout String,
+                                                patternIndex: inout String.Index) {
+        if let currentTextCharacter = currentTextCharacter, !currentTextCharacter.isEmpty {
+            finalText += patternCharactersToBeWritten
+            patternCharactersToBeWritten = ""
+            finalText += currentTextCharacter
+            patternIndex = pattern.index(after: patternIndex)
+        }
+        currentTextIndex = currentText.index(after: currentTextIndex)
+    }
+
+    private func getTextForPatternCharacter(_ patternCharacter: Character, _ text: String) -> String? {
+        switch patternCharacter {
+        case PatternCharacter.lettersAndNumbersPatternCharacter.rawValue:
+            return text
+                .components(separatedBy: CharacterSet(charactersIn: allowedLetters + allowedNumbers).inverted).joined()
+
+        case PatternCharacter.lettersOnlyPatternCharacter.rawValue:
+            return text.components(separatedBy: CharacterSet(charactersIn: allowedLetters).inverted).joined()
+
+        case PatternCharacter.numbersOnlyPatternCharacter.rawValue:
+            return text.components(separatedBy: CharacterSet(charactersIn: allowedNumbers).inverted).joined()
+
+        default:
+            return nil
+        }
+    }
+
+    func getFormatPattern(inputText: String) -> String? {
```

Comment:
You are already passing `getUnformattedText()` to this function why again scrubbing it?

Reply:
I think @ddhar-hw has another call to this method which is not passing getUnformattedText()
@ddhar-hw can we pass unformatted text to this method from setUpTextField method, that way we won't need to do this in this method
`let scrubbedText = getTextForPatternCharacter(PatternCharacter.lettersAndNumbersPatternCharacter.rawValue,
                                                      inputText)`

Reply:
I dont need it - because the other place I am using it is when formatting the text when displaying for the first time

Reply:
noted, removed call to get scrubbed text